sdks/python/apache_beam/utils/windowed_value.py

```diff
@@ -0,0 +1,114 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Core windowing data structures.
+"""
+
+# This module is carefully crafted to have optimal performance when
+# compiled whiel still being valid Python.  Care needs to be taken when
+# editing this file as WindowedValues are created for every element for
+# every step in a Beam pipeline.
+
+#cython: profile=True
+
+from apache_beam.transforms.timeutil import MAX_TIMESTAMP
+from apache_beam.transforms.timeutil import MIN_TIMESTAMP
+from apache_beam.transforms.timeutil import Timestamp
+
+
+class WindowedValue(object):
+  """A windowed value having a value, a timestamp and set of windows.
+
+  Attributes:
+    value: The underlying value of a windowed value.
+    timestamp: Timestamp associated with the value as seconds since Unix epoch.
+    windows: A set (iterable) of window objects for the value. The window
+      object are descendants of the BoundedWindow class.
+  """
+
+  def __init__(self, value, timestamp, windows):
+    # For performance reasons, only timestamp_micros is stored by default
+    # (as a C int). The Timestamp object is created on demand below.
+    self.value = value
+    if isinstance(timestamp, int):
+      self.timestamp_micros = timestamp * 1000000
+    else:
+      self.timestamp_object = (timestamp if isinstance(timestamp, Timestamp)
+                               else Timestamp.of(timestamp))
+      self.timestamp_micros = self.timestamp_object.micros
+    self.windows = windows
+
+  @property
+  def timestamp(self):
+    if self.timestamp_object is None:
+      self.timestamp_object = Timestamp(0, self.timestamp_micros)
+    return self.timestamp_object
+
+  def __repr__(self):
+    return '(%s, %s, %s)' % (
+        repr(self.value),
+        'MIN_TIMESTAMP' if self.timestamp == MIN_TIMESTAMP else
+        'MAX_TIMESTAMP' if self.timestamp == MAX_TIMESTAMP else
+        float(self.timestamp),
+        self.windows)
+
+  def __hash__(self):
+    return hash(self.value) + 3 * self.timestamp_micros + 7 * hash(self.windows)
+
+  # We'd rather implement __eq__, but Cython supports that via __richcmp__
+  # instead.  Fortunately __cmp__ is understood by both (but not by Python 3).
+  def __cmp__(left, right):  # pylint: disable=no-self-argument
+    if type(left) is not type(right):
+      return cmp(type(left), type(right))
+    else:
+      # Don't bother paying the cost of a total ordering.
+      # TODO(robertwb): Avoid the type checks?
+      return not WindowedValue._typed_eq(left, right)
+
+  @staticmethod
+  def _typed_eq(left, right):
+    if (left.timestamp_micros == right.timestamp_micros
+        and left.value == right.value
+        and left.windows == right.windows):
+      return True
+    else:
+      return False
+
+  def with_value(self, new_value):
+    """Creates a new WindowedValue with the same timestamps and windows as this.
+
+    This is the fasted way to create a new WindowedValue.
```

Comment:
Yes, faster than instantiating via the constructor, and faster than calling create() directly (from Python) because that takes three Python arguments. (Python optimizes that path for a single argument.)